core/go/internal/sequencer/originator/state_machine.go

```diff
@@ -0,0 +1,347 @@
+	if txn, ok := o.transactionsByID[txID]; !ok {
+		endorsements := txn.GetEndorsementStatus(ctx)
+		return components.PrivateTxStatus{
+			TxID:         txID.String(),
+			Status:       txn.GetCurrentState().String(),
+			LatestEvent:  txn.GetLatestEvent(),
+			Endorsements: endorsements,
+			Transaction:  txn.PrivateTransaction,
+		}, nil
+	}
+	return components.PrivateTxStatus{
+		TxID:   txID.String(),
+		Status: "unknown",
+	}, nil
+}
+
+// Function evaluateEvent evaluates whether the event is relevant given the current state of the coordinator
+func (o *originator) evaluateEvent(ctx context.Context, event common.Event) (*EventHandler, error) {
+	sm := o.stateMachine
+
+	//Determine if and how this event applies in the current state and which, if any, transition it triggers
+	eventHandlers := stateDefinitionsMap[sm.currentState].Events
+	eventHandler, isHandlerDefined := eventHandlers[event.Type()]
+	if isHandlerDefined {
+		//By default all events in the list are applied unless there is a validator function and it returns false
+		if eventHandler.Validator != nil {
+			valid, err := eventHandler.Validator(ctx, o, event)
+			if err != nil {
+				//This is an unexpected error.  If the event is invalid, the validator should return false and not an error
+				log.L(ctx).Errorf("error validating event %s: %v", event.TypeString(), err)
+				return nil, err
+			}
+			if !valid {
+				// This is perfectly normal sometimes an event happens and is no longer relevant to the transaction so we just ignore it and move on.
+				// We log a warning in case it's not a late-delivered message but something that needs looking in to
+				log.L(ctx).Warnf("originator event %s is not valid for current state %s", event.TypeString(), sm.currentState.String())
+				return nil, nil
+			}
+		}
+		return &eventHandler, nil
+	}
+	return nil, nil
+}
+
+// Function applyEvent updates the internal state of the coordinator with information from the event
+// this happens before the state machine is evaluated for transitions that may be triggered by the event
+// so that any guards on the transition rules can take into account the new internal state of the coordinator after this event has been applied
+func (o *originator) applyEvent(ctx context.Context, event common.Event) error {
+	var err error
+	// First apply the event to the update the internal fine grained state of the coordinator if there is any handler registered for the current state
+	switch event := event.(type) {
+	case *HeartbeatReceivedEvent:
+		err = o.applyHeartbeatReceived(ctx, event)
+	case *TransactionCreatedEvent:
+		err = o.createTransaction(ctx, event.Transaction)
+	case *TransactionConfirmedEvent:
+		err = o.confirmTransaction(ctx, event.From, event.Nonce, event.Hash, event.RevertReason)
+	default:
+		log.L(ctx).Debugf("no action defined for event %T", event)
+
+	}
+	if err != nil {
+		log.L(ctx).Errorf("error applying event %v: %v", event.Type(), err)
+	}
+	return err
+}
```

Comment:
Same again (combined with strong feedback on common-ing up the framework) that we should push this into separate functions in the state machine definition.